rfcs/disallow_intents.md

```diff
@@ -0,0 +1,240 @@
+```
+
+In the above case, the dialogue handler is constructed as a dialogflow named `find_restaurants`.
+
+```python
+@app.dialogue_flow(domain='restaurants', intent='find_restaurants')
+async def find_restaurants(ctx, responder):
+  pass
+
+@find_restaurants.handle(intent='confirm_restaurant_name')
+async def confirm_restaurant(ctx, responder):
+  pass
+```
+
+The `confirm_restaurant_name` intent has the following training queries:
+```
+Burger King
+Mcdonalds
+Chick-fil-A
+...
+```
+
+## Proposed solution
+
+For purely followup intents like `confirm_restaurant_name`, which are defined as intents whose dialogue handlers are
+only associated with a dialogflow and not as an independent dialogue handler, we mask those intents for all predictions
+that are not in the associated dialogflow.
+
+```
+user >> the beatles
+
+[allowed_intents = all intents except `confirm_restaurant_name` since it's associated with find_restaurants df]
+
+bot >> I dont know what that means
+
+user >> find restaurants
+
+[allowed_intents = all intents except `confirm_restaurant_name` since it's associated with find_restaurants df]
+[find_restaurants df is activated now]
+
+bot >> what restaurant
+
+[allowed_intents = all intents and `confirm_restaurant_name` since it's associated with find_restaurants df]
+[find_restaurants df is activated now]
+
+user >> the beatles
+
+bot >> cannot find a restaurant titled "the beatles"
+```
+
+This solution will have no visible API changes.
+
+## Important note:
+
+We have not considered biasing the classifiers in this discussion, as compared to complete blacklisting. Hence, we ignore
+the cases were the true intent was a follow-up but we classified something else. For example:
+
+```
+user >> find a restaurant
+bot >> what restaurant
+user >> oasis place [intent: unknown, true_intent: confirm_restaurant_name]
+bot >> I dont know what that means
+```
+
+NLP biasing is complicated since we need to agree on hyperparams to judge the closeness of NLP predictions and when
+to prefer a lower ranked NLP component prediction to a higher ranked one.
+
+
+## API Design
+
+We have to standardize and refactor the existing `allowed_intents` field since its being used for different purposes like
+masking roles. Here are some design proposals on that:
+
+```python
+nlp.process('I want to move $190 to checkings',
+            allow_intents=['domain.intent'],
+            allow_entities=['domain.intent.entity'],
+            allow_domains=['domain'],
+            allow_roles=['domain.intent.entity.roles'],
+            deny_intents=[],
+            deny_entities=[],
+            deny_domains=[],
+            deny_roles=[]
+        )
+```
+
+I like this design since there isn't much nesting and if users are just using one construct like `allow_intents`, they can just use that
+without using other constructs.
+
+Other designs:
+
+```python
+nlp.process('I want to move $190 to checkings',
+              mask_nlp={
+                'allow': {
```

Comment:
I think it is bit more aesthetic and easy to enumerate if one can specify allowed or denied items through `allow_nlp={...}` or `deny_nlp={...}` arguments, as in the first design. I also like this way of 'allow' and 'deny' as first level keys too, but am wondering if there comes a need to allow/disallow several items at-a-time (and hence it might not be necessary to have a structural hierarchy as in this design!).

Reply:
@murali1996 I was referring to the second design (the allow_domain, allow_intent etc etc). The design doc got updated since :) 

And I agree, the ``allow_nlp`` and ``deny_nlp`` approach is the most concise and reduces developer effort too. Would prefer that as well.

Reply:
@murali1996 Could you expand on `if there comes a need to allow/disallow several items at-a-time` ?

Reply:
I was just wondering if the scenario where we need to populate all the keys 'domains', 'intents', 'entities' and 'roles' at the same time in either 'allow' or 'deny' dicts, is probably rare.